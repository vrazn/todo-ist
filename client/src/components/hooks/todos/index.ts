import useSWR from 'swr';
import useSWRMutation from 'swr/mutation';

import type {
  ITodo,
  ITodoCreationAttributes,
  ITodoDeleteAttributes,
  ITodoUpdateAttributes,
} from 'todo-list';

const headers = {
  'content-type': 'application/json',
};

const GET = async (input: string) => {
  const response = await fetch(input, { headers });
  const json = await response.json();

  if (!response.ok) {
    throw json;
  }
  return json;
};

const POST = async (
  input: string,
  { arg }: { arg: ITodoCreationAttributes },
) => {
  const response = await fetch(input, {
    method: 'POST',
    body: JSON.stringify(arg),
    headers,
  });
  const json = await response.json();

  if (!response.ok) {
    throw json;
  }
  return json;
};

const PATCH = async (
  input: string,
  { arg }: { arg: ITodoUpdateAttributes },
) => {
  const { id, ...body } = arg;

  const response = await fetch(`${input}/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(body),
    headers,
  });
  const json = await response.json();

  if (!response.ok) {
    throw json;
  }
  return json;
};

const DELETE = async (
  input: string,
  { arg }: { arg: ITodoDeleteAttributes },
) => {
  const { id } = arg;

  const response = await fetch(`${input}/${id}`, { method: 'DELETE', headers });
  const json = await response.json();

  if (!response.ok) {
    throw json;
  }
  return json;
};

export const useGetTodos = () => {
  const { data, isLoading, error } = useSWR<ITodo[]>(
    `${process.env.NEXT_PUBLIC_BACKEND_URL}/todos`,
    GET,
    { refreshInterval: 0 },
  );

  return {
    todos: data,
    isLoading,
    error,
  };
};

export const useCreateTodo = () => {
  const { data, trigger, isMutating, error } = useSWRMutation<ITodo>(
    `${process.env.NEXT_PUBLIC_BACKEND_URL}/todos`,
    POST,
  );

  return {
    todo: data,
    trigger,
    isMutating,
    error,
  };
};

export const usePatchTodo = () => {
  const { data, trigger, isMutating, error } = useSWRMutation<ITodo>(
    `${process.env.NEXT_PUBLIC_BACKEND_URL}/todos`,
    PATCH,
  );

  return {
    todo: data,
    trigger,
    isMutating,
    error,
  };
};

export const useDeleteTodo = () => {
  const { data, trigger, isMutating, error } = useSWRMutation<ITodo>(
    `${process.env.NEXT_PUBLIC_BACKEND_URL}/todos`,
    DELETE,
  );

  return {
    todo: data,
    trigger,
    isMutating,
    error,
  };
};
