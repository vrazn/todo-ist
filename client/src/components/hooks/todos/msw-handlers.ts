import { rest } from 'msw';
import { ITodo } from 'todo-list';

export const todos = [
  {
    id: 1,
    title: 'Title 1',
    description: 'Description 1',
    isImportant: true,
    isDone: false,
    dueDate: '2023-01-01',
  },
  {
    id: 2,
    title: 'Title 2',
    description: null,
    isImportant: false,
    isDone: false,
    dueDate: null,
  },
  {
    id: 3,
    title: 'Title 3',
    description: 'Description 3',
    isImportant: true,
    isDone: true,
    dueDate: '2023-01-03',
  },
];

export const todo = todos[0];

export const emptyTodos: ITodo[] = [];

export const error = { message: 'Test error' };

export const getEmpty = rest.get('/todos', (_req, res, ctx) => {
  return res(ctx.delay(100), ctx.status(200), ctx.json(emptyTodos));
});

export const getSuccessStub: ITodo[] = [];

export const getSuccess = rest.get('/todos', (_req, res, ctx) => {
  return res(ctx.delay(100), ctx.status(200), ctx.json(todos));
});

export const getError = rest.get('/todos', (_req, res, ctx) => {
  return res(ctx.delay(100), ctx.status(500), ctx.json(error));
});

export const postSuccess = rest.post('/todos', (_req, res, ctx) => {
  return res(ctx.delay(100), ctx.status(201), ctx.json(todo));
});

export const postError = rest.post('/todos', (_req, res, ctx) => {
  return res(ctx.delay(100), ctx.status(500), ctx.json(error));
});

export const patchSuccess = rest.patch('/todos/:id', (_req, res, ctx) => {
  return res(ctx.delay(100), ctx.status(200), ctx.json(todo));
});

export const patchError = rest.patch('/todos/:id', (_req, res, ctx) => {
  return res(ctx.delay(100), ctx.status(500), ctx.json(error));
});

export const deleteSuccess = rest.delete('/todos/:id', (_req, res, ctx) => {
  return res(ctx.delay(100), ctx.status(200), ctx.json(todo));
});

export const deleteError = rest.delete('/todos/:id', (_req, res, ctx) => {
  return res(ctx.delay(100), ctx.status(500), ctx.json(error));
});
