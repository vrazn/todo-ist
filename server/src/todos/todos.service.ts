import { Injectable, NotFoundException } from '@nestjs/common';
import { snakeCase } from 'change-case';

import { Todo } from './entities/todo.entity';
import { CreateTodoDto } from './dto/create-todo.dto';
import { UpdateTodoDto } from './dto/update-todo.dto';
import { DatabaseTable } from '../database/database.decorator';
import { DatabaseService } from '../database/database.service';

@Injectable()
export class TodosService {
  constructor(
    @DatabaseTable('todos')
    private readonly db: DatabaseService<Todo>,
  ) {}

  async create(createTodoDto: CreateTodoDto): Promise<Todo> {
    const { keys, values, indexes } = Object.keys(createTodoDto).reduce(
      (acc, key, index) => {
        acc.keys.push(snakeCase(key));
        acc.values.push(createTodoDto[key as keyof CreateTodoDto]);
        acc.indexes.push(`$${index + 1}`);
        return acc;
      },
      { keys: [], values: [], indexes: [] } as {
        keys: string[];
        values: string[];
        indexes: string[];
      },
    );

    const [result] = await this.db.insert({
      query: keys.join(', '),
      where: indexes.join(', '),
      variables: values,
    });
    return result;
  }

  async findAll(): Promise<Todo[]> {
    return this.db.select({ query: '*', variables: [] });
  }

  async findOne(id: number): Promise<Todo> {
    const [result] = await this.db.select({
      query: '*',
      where: 'id = $1',
      variables: [id],
    });

    if (!result) throw new NotFoundException();
    return result;
  }

  async update(id: number, updateTodoDto: UpdateTodoDto): Promise<Todo> {
    const keys = Object.keys(updateTodoDto);
    const values = Object.values(updateTodoDto);

    const query = keys
      .map((key, index) => `${snakeCase(key)} = $${index + 1}`)
      .join(', ');

    const [result] = await this.db.update({
      query,
      where: `id = $${keys.length + 1}`,
      variables: [...values, id],
    });

    return result;
  }

  async remove(id: number): Promise<Todo> {
    const [result] = await this.db.delete({
      where: 'id = $1',
      variables: [id],
    });

    return result;
  }
}
