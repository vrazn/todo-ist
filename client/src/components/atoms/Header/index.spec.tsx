import { render, waitFor } from '@testing-library/react';
import { setupServer } from 'msw/node';
import { SWRConfig } from 'swr';

import * as handlers from '@/components/hooks/todos/msw-handlers';

import { Header, getProgressWidth } from './index';

const server = setupServer();

const TestComponent = () => (
  <SWRConfig value={{ dedupingInterval: 0, provider: () => new Map() }}>
    <Header />
  </SWRConfig>
);

describe('progress width', () => {
  it('returns various widths depending on the screen size', () => {
    expect(
      getProgressWidth({
        xs: false,
        sm: false,
        md: false,
        lg: false,
        xl: false,
        xxl: true,
      }),
    ).toBe(300);
    expect(
      getProgressWidth({
        xs: false,
        sm: false,
        md: false,
        lg: false,
        xl: true,
        xxl: false,
      }),
    ).toBe(300);
    expect(
      getProgressWidth({
        xs: false,
        sm: false,
        md: false,
        lg: true,
        xl: false,
        xxl: false,
      }),
    ).toBe(250);
    expect(
      getProgressWidth({
        xs: false,
        sm: false,
        md: true,
        lg: false,
        xl: false,
        xxl: false,
      }),
    ).toBe(250);
    expect(
      getProgressWidth({
        xs: false,
        sm: true,
        md: false,
        lg: false,
        xl: false,
        xxl: false,
      }),
    ).toBe(200);
    expect(
      getProgressWidth({
        xs: true,
        sm: false,
        md: false,
        lg: false,
        xl: false,
        xxl: false,
      }),
    ).toBe(100);
  });
});

describe('Header', () => {
  beforeAll(() => server.listen());
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());

  describe('on initial load', () => {
    beforeEach(() => server.use(handlers.getSuccess));

    it('renders without change', () => {
      const { container } = render(<TestComponent />);
      expect(container).toMatchSnapshot();
    });

    it('renders avatar only while the data is being loaded', () => {
      const { getByRole, queryByRole } = render(<TestComponent />);

      expect(getByRole('img')).toBeInTheDocument();
      expect(queryByRole('progressbar')).not.toBeInTheDocument();
    });
  });

  describe('on fetch returning an empty array', () => {
    beforeEach(() => server.use(handlers.getEmpty));

    it('renders avatar and an empty progress bar after loading is complete', async () => {
      const { findByRole, findByText } = render(<TestComponent />);

      await waitFor(async () => {
        expect(await findByRole('img')).toBeInTheDocument();
        const progressbar = await findByRole('progressbar');
        expect(progressbar).toBeInTheDocument();
        expect(await findByText('0%')).toBeInTheDocument();
      });
    });
  });

  describe('on fetch returning non-empty array', () => {
    beforeEach(() => server.use(handlers.getSuccess));

    it('renders without change', () => {
      const { container } = render(<TestComponent />);
      expect(container).toMatchSnapshot();
    });

    it('renders avatar only while the data is being loaded', () => {
      const { getByRole, queryByRole } = render(<TestComponent />);

      expect(getByRole('img')).toBeInTheDocument();
      expect(queryByRole('progressbar')).not.toBeInTheDocument();
    });

    it('renders avatar and fills the progress bar after loading is complete', async () => {
      const { findByRole, findByText } = render(<TestComponent />);

      await waitFor(async () => {
        expect(await findByRole('img')).toBeInTheDocument();
        const progressbar = await findByRole('progressbar');
        expect(progressbar).toBeInTheDocument();

        const doneCount = handlers.todos.reduce((acc, curr) => {
          if (curr.isDone) acc++;
          return acc;
        }, 0);

        const percent = Math.round((doneCount / handlers.todos.length) * 100);

        expect(await findByText(`${percent}%`)).toBeInTheDocument();
      });
    });
  });

  describe('on fetch returning error', () => {
    beforeEach(() => server.use(handlers.getError));

    it('renders without change', () => {
      const { container } = render(<TestComponent />);
      expect(container).toMatchSnapshot();
    });

    it('renders avatar and an empty progress bar after loading is complete', async () => {
      const { findByRole, findByText } = render(<TestComponent />);

      await waitFor(async () => {
        expect(await findByRole('img')).toBeInTheDocument();
        const progressbar = await findByRole('progressbar');
        expect(progressbar).toBeInTheDocument();
        expect(await findByText('0%')).toBeInTheDocument();
      });
    });
  });
});
