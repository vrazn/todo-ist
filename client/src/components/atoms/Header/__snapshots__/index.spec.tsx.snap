// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header on fetch returning error renders without change 1`] = `
<div>
  <header
    class="ant-layout-header header"
  >
    <div
      class="ant-row ant-row-center css-dev-only-do-not-override-ixblex"
    >
      <div
        class="ant-col ant-col-xs-23 ant-col-sm-23 ant-col-md-21 ant-col-lg-20 ant-col-xl-18 css-dev-only-do-not-override-ixblex"
      >
        <div
          class="ant-row ant-row-end ant-row-middle css-dev-only-do-not-override-ixblex"
        >
          <div
            class="ant-space css-dev-only-do-not-override-ixblex ant-space-horizontal ant-space-align-center"
          >
            <div
              class="ant-space-item"
            >
              <span
                class="ant-avatar ant-avatar-circle ant-avatar-image css-dev-only-do-not-override-ixblex"
                style="width: 64px; height: 64px; line-height: 64px; font-size: 18px;"
              >
                <img
                  src="https://avatars.githubusercontent.com/u/8733843?s=96"
                />
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </header>
</div>
`;

exports[`Header on fetch returning non-empty array renders without change 1`] = `
<div>
  <header
    class="ant-layout-header header"
  >
    <div
      class="ant-row ant-row-center css-dev-only-do-not-override-ixblex"
    >
      <div
        class="ant-col ant-col-xs-23 ant-col-sm-23 ant-col-md-21 ant-col-lg-20 ant-col-xl-18 css-dev-only-do-not-override-ixblex"
      >
        <div
          class="ant-row ant-row-end ant-row-middle css-dev-only-do-not-override-ixblex"
        >
          <div
            class="ant-space css-dev-only-do-not-override-ixblex ant-space-horizontal ant-space-align-center"
          >
            <div
              class="ant-space-item"
            >
              <span
                class="ant-avatar ant-avatar-circle ant-avatar-image css-dev-only-do-not-override-ixblex"
                style="width: 64px; height: 64px; line-height: 64px; font-size: 18px;"
              >
                <img
                  src="https://avatars.githubusercontent.com/u/8733843?s=96"
                />
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </header>
</div>
`;

exports[`Header on initial load renders without change 1`] = `
<div>
  <header
    class="ant-layout-header header"
  >
    <div
      class="ant-row ant-row-center css-dev-only-do-not-override-ixblex"
    >
      <div
        class="ant-col ant-col-xs-23 ant-col-sm-23 ant-col-md-21 ant-col-lg-20 ant-col-xl-18 css-dev-only-do-not-override-ixblex"
      >
        <div
          class="ant-row ant-row-end ant-row-middle css-dev-only-do-not-override-ixblex"
        >
          <div
            class="ant-space css-dev-only-do-not-override-ixblex ant-space-horizontal ant-space-align-center"
          >
            <div
              class="ant-space-item"
            >
              <span
                class="ant-avatar ant-avatar-circle ant-avatar-image css-dev-only-do-not-override-ixblex"
                style="width: 64px; height: 64px; line-height: 64px; font-size: 18px;"
              >
                <img
                  src="https://avatars.githubusercontent.com/u/8733843?s=96"
                />
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </header>
</div>
`;
